pulsar-io/kinesis/src/main/java/org/apache/pulsar/io/kinesis/KinesisSource.java

```diff
@@ -80,31 +85,58 @@ public void open(Map<String, Object> config, SourceContext sourceContext) throws
         queue = new LinkedBlockingQueue<KinesisRecord> (kinesisSourceConfig.getReceiveQueueSize());
         workerId = InetAddress.getLocalHost().getCanonicalHostName() + ":" + UUID.randomUUID();
         
-        AWSCredentialsProvider credentialsProvider = createCredentialProvider(
+        AwsCredentialProviderPlugin credentialsProvider = createCredentialProvider(
                 kinesisSourceConfig.getAwsCredentialPluginName(), 
                 kinesisSourceConfig.getAwsCredentialPluginParam());
-        
-        kinesisClientLibConfig = 
-                new KinesisClientLibConfiguration(kinesisSourceConfig.getApplicationName(),
-                        kinesisSourceConfig.getAwsKinesisStreamName(),
-                        credentialsProvider,
-                        workerId)
-                .withRegionName(kinesisSourceConfig.getAwsRegion())
-                .withInitialPositionInStream(kinesisSourceConfig.getInitialPositionInStream());
-        
-        if (kinesisSourceConfig.getInitialPositionInStream() == InitialPositionInStream.AT_TIMESTAMP) {
-           kinesisClientLibConfig.withTimestampAtInitialPositionInStream(kinesisSourceConfig.getStartAtTime());
-        }
-        
+
+        KinesisAsyncClient kClient = kinesisSourceConfig.buildKinesisAsyncClient(credentialsProvider);
         recordProcessorFactory = new KinesisRecordProcessorFactory(queue, kinesisSourceConfig);
-        
-        worker = new Worker(recordProcessorFactory, kinesisClientLibConfig);
-        worker.run();
+        configsBuilder = new ConfigsBuilder(kinesisSourceConfig.getAwsKinesisStreamName(),
+                                            kinesisSourceConfig.getApplicationName(),
+                                            kClient,
+                                            kinesisSourceConfig.buildDynamoAsyncClient(credentialsProvider),
+                                            kinesisSourceConfig.buildCloudwatchAsyncClient(credentialsProvider),
+                                            workerId,
+                                            recordProcessorFactory);
+
+        RetrievalConfig retrievalConfig = configsBuilder.retrievalConfig();
+        if (!kinesisSourceConfig.isUseEnhancedFanOut()) {
+            retrievalConfig.retrievalSpecificConfig(
+                    new PollingConfig(kinesisSourceConfig.getAwsKinesisStreamName(),
+                                      kClient));
+        }
+
+        retrievalConfig.initialPositionInStreamExtended(kinesisSourceConfig.getStreamStartPosition());
+
+        scheduler = new Scheduler(
```

Comment:
I don't remember seeing any of this scheduler / threading stuff in their code. They must not have ever tested it.

Reply:
the scheduler is a new thing introduced in v2 of the KCL, but there was an analogous thing called a "worker". They failed to start the worker in another thread which was why it didn't work :P